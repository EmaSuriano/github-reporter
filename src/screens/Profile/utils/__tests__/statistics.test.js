import {
  createConfiguration,
  createData,
  getEdges,
  getEndCursor,
  getPageInfo,
  getStatistics,
  getTotalCount,
  hasNextPage,
  updateRepositoryData
} from "../statistics";

import edge from "./fixtures/edge";
import edgeForked from "./fixtures/edgeForked";

describe("Statistics", () => {
  describe("createConfiguration", () => {
    let result;

    beforeEach(() => {
      const user = "wesbos";
      const configuration = { width: 100, height: 100 };
      result = createConfiguration(user, configuration);
    });

    it("should match configuration object with the snapshot", () =>
      expect(result).toMatchSnapshot());

    describe("onClick instance", () => {
      let onClickInstance;

      const spyWindow = jest.spyOn(window, "open");

      describe('when canvas is equal to "by-language"', () => {
        beforeEach(() => {
          const user = "wesbos";
          const data = [
            {
              _model: { label: "linkLabel" },
              _chart: { canvas: { id: "by-language" } }
            }
          ];

          result.options.onClick(user, data);
        });

        afterEach(() => {
          jest.clearAllMocks();
        });

        it("should call window open once", () =>
          expect(spyWindow).toHaveBeenCalledTimes(1));

        it("should call window open with the proper link and _blank property", () =>
          expect(spyWindow).toHaveBeenCalledWith(
            "https://github.com/wesbos?utf8=%E2%9C%93&tab=repositories&q=&type=source&language=linkLabel",
            "_blank"
          ));
      });

      describe('when canvas is equal to "by-repository"', () => {
        beforeEach(() => {
          const user = "wesbos";
          const data = [
            {
              _model: { label: "linkLabel" },
              _chart: { canvas: { id: "by-repository" } }
            }
          ];

          result.options.onClick(user, data);
        });

        afterEach(() => {
          jest.clearAllMocks();
        });

        it("should call window open once", () =>
          expect(spyWindow).toHaveBeenCalledTimes(1));

        it("should call window open with the proper link and _blank property", () =>
          expect(spyWindow).toHaveBeenCalledWith(
            "https://github.com/wesbos/linkLabel",
            "_blank"
          ));
      });
    });
  });

  describe("createData", () => {
    let result;

    beforeEach(() => {
      const data = { languages: { HTML: { stars: 100 } } };
      const from = "languages";
      const currentProp = "stars";
      result = createData(data, from, currentProp);
    });

    it("should match the object created by the function with the datasets and labels", () =>
      expect(result).toMatchSnapshot());
  });

  describe("getEdges", () => {
    let result;

    beforeEach(() => {
      const statistics = { user: { repositories: { edges: [5, 2] } } };
      result = getEdges(statistics);
    });

    it('should return edges "[5, 2]"', () => expect(result).toEqual([5, 2]));
  });

  describe("getEndCursor", () => {
    let result;

    beforeEach(() => {
      const testData = { endCursor: "cursorId" };
      result = getEndCursor(testData);
    });

    it('should return "cursorId" as endCursor prop', () =>
      expect(result).toBe("cursorId"));
  });

  describe("getPageInfo", () => {
    let result;

    beforeEach(() => {
      const statistics = { user: { repositories: { pageInfo: "pageInfo" } } };
      result = getPageInfo(statistics);
    });

    it('should return "pageInfo"', () => expect(result).toBe("pageInfo"));
  });

  describe("getStatistics", () => {
    let result;

    describe("when edge has forked", () => {
      beforeEach(() => {
        result = getStatistics([], edgeForked);
      });

      it("should return respositories property without commits and stars", () =>
        expect(result.repositories).toEqual({ "anchor-css-transition": {} }));

      it("should return languages property without repositories and stars", () =>
        expect(result.languages).toEqual({ Unknown: { commits: 1 } }));
    });

    describe("when edge has not forked", () => {
      beforeEach(() => {
        result = getStatistics([], edge);
      });

      it("should return", () =>
        expect(result).toEqual({
          languages: {
            JavaScript: { commits: 47, repositories: 1, stars: 24 }
          },
          repositories: { "gatsby-starter-mate": { commits: 47, stars: 24 } }
        }));
    });
  });

  describe("getTotalCount", () => {
    let result;

    beforeEach(() => {
      const testData = { totalCount: 5 };
      result = getTotalCount(testData);
    });

    it('should return "5" as totalCount', () => expect(result).toBe(5));
  });

  describe("hasNextPage", () => {
    let result;

    describe('when "hasNextPage" prop exists', () => {
      beforeEach(() => {
        const testData = { hasNextPage: true };
        result = hasNextPage(testData);
      });

      it("should return true", () => expect(result).toBeTruthy());
    });

    describe('when "hasNextPage" prop does not exist', () => {
      beforeEach(() => {
        result = hasNextPage({});
      });

      it("should return false as default value", () =>
        expect(result).toBeFalsy());
    });
  });

  describe("updateRepositoryData", () => {
    let result;

    beforeEach(() => {
      const repositoryInfo = {
        user: {
          repositories: {
            pageInfo: {
              hasNextPage: true,
              endCursor: ""
            },
            edges: []
          }
        }
      };

      const edges = [
        {
          node: {
            name: "JavaScriptRepo",
            isFork: false,
            primaryLanguage: {
              name: "JavaScript"
            }
          }
        }
      ];
      const pageInfo = { hasNextPage: false, endCursor: "endCursor" };
      result = updateRepositoryData(repositoryInfo, edges, pageInfo);
    });

    it("should have one edge with 'JavaScriptRepo' as name", () => {
      expect(result.user.repositories.edges).toEqual([
        {
          node: {
            name: "JavaScriptRepo",
            isFork: false,
            primaryLanguage: { name: "JavaScript" }
          }
        }
      ]);
    });
  });
});
