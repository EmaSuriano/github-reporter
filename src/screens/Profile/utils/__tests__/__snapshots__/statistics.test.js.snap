// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Statistics createData should match the object created by the function with the datasets and labels 1`] = `
Object {
  "datasets": Array [
    Object {
      "backgroundColor": Array [
        "#A4BDFC",
        "#7AE7BF",
        "#DBADFF",
        "#FF887C",
        "#FBD75B",
        "#FFB878",
        "#46D6DB",
        "#E1E1E1",
        "#5484ED",
        "#51B749",
        "#DC2127",
      ],
      "data": Array [
        100,
      ],
    },
  ],
  "labels": Array [
    "HTML",
  ],
}
`;
